### PR TITLE
Make recording of a session optional - Fix #1163

### LINKED ISSUE #1163: [Feature Request] Make recording of a session optional
When starting a BBB session via greenlight the API call to BBB to create the conference always contains `record=true`. 

Since we mostly use our BBB for office meetings recording by default is not necessary and causes the disk to fill up pretty quickly. 

It would be nice to have a room option like "Allow recording of sessions" which toggles if `record=true` will be passed to BBB on creation. 

As far as I can see right now this option is hardcoded in the file `app/controllers/concerns/joiner.rb` (Line 89).

### PR DESCRIPTION
**Pull request by yanosz**
Fix #1163 by making recording configurable

By this PR, `recording` is among the room-features configured in the sample env. Due to its privacy implications, it has to be enabled explicitly. In addition, the administator can configure recording-options using the frontend. _Unless enabled explicitly, rooms are not recoreded._ **Note: This changes default behaviour.**

If enabled, users can configure rooms to record session. This flag is set on a per-room basis and cannot be changed while the room is active (i.e. a session is running)

### CODE REVIEW

**Comment by Arnei** [comment]
Hello,
Is this sketch already intended to be working? I ask because I tried this pull request, but it does not work for me. To be more specific, the new "Recording" room setting does not seem to have an effect on whether recording files are created by bigbluebutton or not (for example, files are created in `/var/bigbluebutton/recording/raw ` no matter the setting). Or did I get something wrong?

**Comment by yanosz** [comment]
> Hello,
> Is this sketch already intended to be working? I ask because I tried this pull request, but it does not work for me. To be more specific, the new "Recording" room setting does not seem to have an effect on whether recording files are created by bigbluebutton or not (for example, files are created in `/var/bigbluebutton/recording/raw ` no matter the setting). Or did I get something wrong?

Heiho,

thanks for feedback and testing! Initially, I planed to do some testing, based on changes caused by #1222 and revisit this at this coming point in time. Basically, this patch was about prototyping the UX regarding the greenlight's architecture. It's my first PR for greenlight, anyway :-).

Nevertheless, I can do some debugging in case sbd. (@Arnei you?) is interested at this patch as of now.

**Comment by yanosz** [comment]
ok, CI is happy, parameters should be passed

**Comment by Arnei** [comment]
Ah, makes sense that it wouldn't prevent recording if it was just a UX prototype :D.

I've tried the newest commit, as it is labeled "pass parameter on creation", and now bigbluebutton won't create any archive files under `/var/bigbluebutton/recording/raw`no matter if the recording option was set to true or false on creation. 

And thank you for working on this!

**Comment by zner0L** [comment]
@yanosz or anyone, can you tell me what's missing from this to be merged? I really need to be able to disable recordings, but if possible I would like to wait for a release before using these changes, but if it takes long I'll just patch it myself…

**Comment by yanosz** [comment]
@zner0L Hard to say. There isn't much feedback from the maintainers, yet. I guess, that a fair amount of new comitters rushed into this project in the last days and things could take time.

I'd helpfult is you coud test the branch as of this PR, as of now and provide some feedback.

**Comment by farhatahmad** [comment]
Hi @yanosz,

Thanks for submitting the PR. The reason I haven't replied is because I'm still unsure with exactly how this should work. We haven't had the time internally to discuss this yet so I can't quite give a deadline on when this will be merged

**Comment by yanosz** [comment]
Hei @farhatahmad 
> 
> Thanks for submitting the PR. The reason I haven't replied is because I'm still unsure with exactly how this should work. We haven't had the time internally to discuss this yet so I can't quite give a deadline on when this will be merged

Thanks your time and answer! I'm also unsure, with exactly how this should work. In essence, greenlight promotes custimization, whereby regional peculiarities (such as the GDBR)  can be with implemented very low effort. Nevertheless, the demand in #1163 appears to be sound. Also, I'm thinking, hat neither the architecture of greenlight, nor the UX for both users and and nor the complexity for customizers can reasonable address evovled, but local requirements such as GDPR-compliance, for certain. Should greenlight - as a project - implement various business use-cases and requirements or should it stay a minimalist appraoch to a bbb frontend, whereby other frontends exists for diverse use-cases? I'm not certain.

That being said, I'm also not certain, if GDPR-compliance actuallually justifies a fork of greenlight.

For the moment, it appears to reasonable to me to maintain this branch of greenlight and ask for testers.

**Comment by gilou** [comment]
I might give it a go indeed, for legal and technical reasons, not enabling the recording by default is very important!

**Comment by Arnei** [comment]
Seems to be working mighty fine so far. Thanks again!

Things I found:
1. Recording_SessionActive
"Session Active" is displayed on room creation. It's not doing any harm, just confusing. Otherwise, "Session Active" is displayed correctly.

**Comment by yanosz** [comment]
@Arnei Thanks for your feedback. I'll have a look at it this weekend.

**Comment by Arnei** [comment]
Got another bug report:

Usually, when you try to join a conference that has not started yet, you get a message looking like so (I hope the german has no detrimental effect on the explanatory power of the image):
ConferenceHasntStartedYet

If I try to do the same thing in a room with recording enabled, after consenting, entering my name and clicking join, the consent button is displayed again:
ConferenceWaitError

The consent button will then stop doing anything useful, while the message "the conference hasn't started yet" won't ever be displayed.

**Comment by yanosz** [comment]
@Arnei thanks for your feedback. I guess, that this is an interesting side effect of unstable DOM manipulations.

I think I'll ditch the cosent-button as explained in #1163  and have GDPR-related consent issues in #1619

**Comment by yanosz** [comment]
I think, that this feature is ready for merged (sorry, for not marking this PR as draf before, I just discovered this functionality).

i'd be good to have some addtional testing. To the best of my knowledge, there is no test automation for both the modal dialogs and the joiner aspect.

@basisbit: Ping.

**Comment by basisbit** [comment]
@yanosz did you already sign the [Contributor License Agreement ](https://docs.bigbluebutton.org/support/faq.html#why-do-i-need-to-sign-a-contributor-license-agreement-to-contribute-source-code)and send it to Fred by Email?

**Comment by yanosz** [comment]
@basisbit - not yet. My last PR for the docs was merged nevertheless. 
Also, this PR contains code from @ichdasich.

**Comment by basisbit** [comment]
as long as you sign the CLA now, it should be fine

**Comment by yanosz** [comment]
well... I've to admit that I'm somewhat afraid to sign legally binding documents, that I don't fully understand. I'm not used to U.S. Federal law, I'm not an U.S. citizen and I don't fully understand the legal implications of all details. Probably I need some days to even understand what I'm signing.

**Comment by farhatahmad** [comment]
@yanosz CLA is only needed for BigBlueButton not Greenlight

**Comment by yanosz** [comment]
> @yanosz CLA is only needed for BigBlueButton not Greenlight

Thanks for the clarificiation! 

Signing a CLA is not a problem per-se, but I'd like to contact some FSFE related lawers beforehand, then. As said, that would take some days. Please let me know if it's needed nevertheless.

**Comment by ichdasich** [comment]
Awesome; No CLA works for me.

**Comment by matiasilva** [comment]
Hello! Just a quick question - I want to run this in my setup, how can I do so? Also, how did you develop this? Did you run the Rails application locally? If so, could you perhaps explain how?

**Comment by yanosz** [comment]
Hei,
> Hello! Just a quick question - I want to run this in my setup and I believe I have managed to merge it into my repository. Unfortunately, after rebuilding the image, nothing seems to happen.. Also, how did you develop this? Did you run the Rails application locally? If so, could you perhaps explain how?

May we move on to on https://groups.google.com/d/forum/bigbluebutton-dev?
I guess, there's could be sth. wrong in your docker setup, that is unrelated to this PR.

**Comment by matiasilva** [comment]
I took your first approach before even asking here and could confirm that the code was there and building. It turned out that `recording` was not added in the room features, dumb me! It works like a charm, excellent. Perhaps you could chime in on my topic on the Google group? Thanks for the great work!

**Comment by yanosz** [comment]
@matiasilva thx for your feedback and testing this PR. I'll try to answer your questions on bbb-dev.

**Comment by Cistoge** [comment]
Attached is a version of the patch that applies cleanly to release-2.6.1

[recording.patch.txt](https://github.com/bigbluebutton/greenlight/files/4696321/recording.patch.txt)

**Comment by yanosz** [comment]
I don't see any relevant conflicts - it's just misplaced whitespace - IMHO.
I'd like to wait for a review / timeline before a keep on working.

**Comment by Cistoge** [comment]
I had already customized greenlight 2.5.6 with your patch and now adopted it for a customized greenlight 2.6.1;  that is why I published the new version here, so that the work does not need to be done twice.

**Comment by basisbit** [comment]
@farhatahmad any chance to get this reviewed and merged soon-ish? About all BBB server providers who use Greenliight in the EU are currently waiting for this to get merged (because of legal requirements) and so far have to manually patch every new release of greenlight.

**Comment by yanosz** [comment]
Sorry for the spam ... I'was just trying to align the whitespace, when origins got messed up. It should be fine, again.

**Comment by basisbit** [comment]
yet another week without any reaction, not even a comment. *nudge* @farhatahmad ?

**Comment by farhatahmad** [comment]
@basisbit No need for the attitude :). My comment is the same as it's been since this topic was first brought up. You were also on the call last week when I mentioned that I plan to fully review and merge it into the next version of Greenlight (`v2.7`). 

I've been focused on making `v2.6.x` stable due to the big changes that were made in that version. Would you prefer I leave the bugs and focus solely on this issue? 

@yanosz I appreciate all the work you and everyone else put in to get this PR to the state it is now. I still plan on merging this once I can confirm the `v2.6.x` is stable enough so that I can begin working on `v2.7`

**Comment by matiasilva** [comment]
Great work, as always :)

File: config/application.rb [hunk 1]
```diff
@@ -52,16 +52,16 @@ class Application < Rails::Application
 
     # Use standalone BigBlueButton server.
     config.bigbluebutton_endpoint = if ENV["BIGBLUEBUTTON_ENDPOINT"].present?
-       ENV["BIGBLUEBUTTON_ENDPOINT"]
```
Comment (farhatahmad) [inline config/application.rb:55]: Please exclude this file from the PR

File: config/application.rb [hunk 2]
```diff
@@ -144,12 +144,12 @@ class Application < Rails::Application
 
     # Default registration method if the user does not specify one
     config.registration_method_default = if ENV["DEFAULT_REGISTRATION"] == "invite"
-      config.registration_methods[:invite]
```
Comment (farhatahmad) [inline config/application.rb:147]: Same as above

File: sample.env [hunk 3]
```diff
@@ -150,7 +150,8 @@ RELATIVE_URL_ROOT=/b
 #   require-moderator-approval: Require moderators to approve new users before they can join the room
 #   anyone-can-start: Allows anyone with the join url to start the room in BigBlueButton
 #   all-join-moderator: All users join as moderators in BigBlueButton
-ROOM_FEATURES=mute-on-join,require-moderator-approval,anyone-can-start,all-join-moderator
+#   recording: Sessions are recorded
+ROOM_FEATURES=mute-on-join,require-moderator-approval,anyone-can-start,all-join-moderator,recording
```
Comment (farhatahmad) [inline sample.env:154]: No need to add it here since this was all moved to the Room Configuration tab

**Comment by ichdasich** [comment]
I think the last item is for enabling that setting in the room configuration tab, no?

File: sample.env (hunk 3 above)
Comment (farhatahmad) [reply sample.env:154]: Ahh yea @ichdasich is correct. You can ignore this

File: config/application.rb (hunk 1 above)
Comment (yanosz) [reply config/application.rb:55]: thx for the review - unfortunatly, I'm very busy, atm.
The line is there to make rubocop happy, by fixing identation. i'll take care of it by mid of July.

Comment (farhatahmad) [reply config/application.rb:55]: No worries. I'll just merge it in and fix it myself. Thanks!

**Comment by farhatahmad** [comment]
Made some small changes/improvements in #1851. Feel free to let me know what you guys think :)

**Comment by sondos1992** [comment]
Hi
I've a question can i add automatic recording and download feature to BBB .

**Comment by basisbit** [comment]
@sondos1992 for support requests / such questions, please ask at the mailing list: https://groups.google.com/forum/#!forum/bigbluebutton-users

**Comment by sondos1992** [comment]
it keep showing error #254 connection to server can't join

**Comment by basisbit** [comment]
@sondos1992 for support requests / such questions, please ask at the mailing list: https://groups.google.com/forum/#!forum/bigbluebutton-users

**Pull request merged**